hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/AltoChannelConnector.java

```diff
@@ -0,0 +1,215 @@
+                new byte[UUID_SIZE_IN_BYTES], UNFRAGMENTED_MESSAGE);
+        encodeUUID(initialFrame.content, 0, clientUuid);
+        clientUuidMessage.add(initialFrame);
+        if (!channel.write(clientUuidMessage)) {
+            throw new HazelcastException("Cannot write authentication bytes to the channel");
+        }
+    }
+
+    private void onSuccessfulChannelConnection(Channel channel, int index) {
+        synchronized (altoChannels) {
+            if (connectionFailed()) {
+                // It might be the case that the connection or any
+                // of the channels are failed after this channel
+                // is established. We need to close this one as well
+                // to not leak any channels.
+                logger.warning("Closing the Alto channel " + channel + " for "
+                        + connection.toLessDescriptiveString()
+                        + " as one of the connections are failed.");
+                onFailure(channel);
+                return;
+            }
+
+            altoChannels[index] = channel;
+        }
+
+        logger.info("Successfully connected to Alto channel " + channel + " for "
+                + connection.toLessDescriptiveString());
+
+        if (remaining.decrementAndGet() == 0) {
+            connection.setAltoChannels(altoChannels);
+
+            // If the connection is alive at this point, but
+            // closes afterward, the channels will be cleaned up
+            // properly in the connection's close method, because
+            // we have already written the channels.
+
+            // If the connection is not alive at this point, the channels
+            // might or might not be closed, depending on the order of the
+            // close and setAltoChannels calls. We will close channels
+            // if the connection is not alive here, just in case, as it is
+            // OK to call close on already closed channels.
+            if (!connection.isAlive()) {
+                logger.warning("Closing all Alto channel connections for "
+                        + connection.toLessDescriptiveString() + " as the connection is closed.");
+                closeAllChannels();
+            } else {
+                logger.info("All Alto channel connections are established for the "
+                        + connection.toLessDescriptiveString());
+            }
+        }
+    }
+
+    private void onFailure(Channel channel) {
+        synchronized (altoChannels) {
+            closeChannel(channel);
+            if (failed) {
+                return;
+            }
+
+            failed = true;
+            closeAllChannels();
+        }
+    }
+
+    private boolean connectionFailed() {
+        return failed || !connection.isAlive();
+    }
+
+    private void closeChannel(Channel channel) {
+        if (channel == null) {
+            return;
+        }
+
+        try {
+            channel.close();
```

Comment:
Not directly related with this PR but `channel.close` does not set the closedTimestamp as we set for client connections, and there can be similar closed timestamp on the channels as well. Why did we choose a different approach and used and atomicInteger CLOSED to do it? Why not have the same approach and the timestamps always help for troubleshooting.

Reply:
I didn't implement that one, so I don't know. But, I assume it is the simplest thing that does the job, and the added information from the close time was not needed until this point, as we don't have it either here or in the TcpServerConnection